af/fastpath/fastpath/portable_queue.py

```diff
@@ -0,0 +1,71 @@
+import six
+import platform
+import multiprocessing
+from multiprocessing.queues import Queue as BaseQueue
+
+# taken from: https://github.com/binux/pyspider/blob/d49605d6d1f07d325ff7a4ca311dc5405ed93159/pyspider/libs/multiprocessing_queue.py
+
+class SharedCounter(object):
+    """ A synchronized shared counter.
+    The locking done by multiprocessing.Value ensures that only a single
+    process or thread may read or write the in-memory ctypes object. However,
+    in order to do n += 1, Python performs a read followed by a write, so a
+    second process may read the old value before the new one is written by the
+    first process. The solution is to use a multiprocessing.Lock to guarantee
+    the atomicity of the modifications to Value.
+    This class comes almost entirely from Eli Bendersky's blog:
+    http://eli.thegreenplace.net/2012/01/04/shared-counter-with-pythons-multiprocessing/
+    """
+
+    def __init__(self, n=0):
+        self.count = multiprocessing.Value('i', n)
+
+    def increment(self, n=1):
+        """ Increment the counter by n (default = 1) """
+        with self.count.get_lock():
```

Comment:
A whole dedicated class to call get_lock() seems a bit bloated. I would do the locking in the Queue class. Also it would allow decrementing the value directly instead of using "increment(-1)"

Reply:
Yeah I guess so. I figured I would steal some already tested and written code though rather than writing it myself, but I am also ok with simplifying it.